src/e2e_test/perf_tool/pkg/metrics/prometheus_recorder.go

```diff
@@ -0,0 +1,290 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/portforward"
+	"k8s.io/client-go/transport/spdy"
+
+	"px.dev/pixie/src/e2e_test/perf_tool/experimentpb"
+	"px.dev/pixie/src/e2e_test/perf_tool/pkg/cluster"
+)
+
+type prometheusRecorderImpl struct {
+	clusterCtx *cluster.Context
+	spec       *experimentpb.PrometheusScrapeSpec
+	eg         *errgroup.Group
+	resultCh   chan<- *ResultRow
+
+	wg     sync.WaitGroup
+	stopCh chan struct{}
+	fws    []*forwarder
+}
+
+func (r *prometheusRecorderImpl) Start(ctx context.Context) error {
+	if err := r.startPortForward(); err != nil {
+		return err
+	}
+	r.stopCh = make(chan struct{})
+	r.wg.Add(1)
+	r.eg.Go(func() error {
+		defer r.wg.Done()
+		err := r.run()
+		if err != nil {
+			err = fmt.Errorf("error recording prometheus metrics: %w", err)
+		}
+		return err
+	})
+	return nil
+}
+
+func (r *prometheusRecorderImpl) Close() {
+	if r.stopCh != nil {
+		close(r.stopCh)
+		r.wg.Wait()
+		r.stopCh = nil
+	}
+	for _, fw := range r.fws {
+		fw.Close()
+	}
+}
+
+func (r *prometheusRecorderImpl) run() error {
+	d, err := types.DurationFromProto(r.spec.ScrapePeriod)
+	if err != nil {
+		return err
+	}
+	t := time.NewTicker(d)
+	for {
+		select {
+		case <-r.stopCh:
+			return nil
+		case <-t.C:
+			if err := r.scrape(); err != nil {
+				return err
+			}
+		}
+	}
+}
+
+func (r *prometheusRecorderImpl) scrape() error {
+	eg := errgroup.Group{}
+	for _, fw := range r.fws {
+		eg.Go(r.scrapeFunc(fw))
+	}
+	return eg.Wait()
+}
+
+func (r *prometheusRecorderImpl) scrapeFunc(fw *forwarder) func() error {
+	op := func() error {
+		resp, err := http.Get(fmt.Sprintf("http://localhost:%d/metrics", fw.localPort))
+		if err != nil {
+			return err
+		}
+		ts := time.Now()
+		metrics, err := (&expfmt.TextParser{}).TextToMetricFamilies(resp.Body)
+		if err != nil {
+			return backoff.Permanent(err)
+		}
+		for _, mf := range metrics {
+			name, ok := r.spec.MetricNames[mf.GetName()]
+			if !ok {
+				// Only collect metrics named in the spec.
+				continue
+			}
+			// Only collect counters and gauges for now.
+			if !(mf.GetType() == io_prometheus_client.MetricType_COUNTER || mf.GetType() == io_prometheus_client.MetricType_GAUGE) {
+				continue
+			}
+			for _, m := range mf.GetMetric() {
+				row := newResultRow()
+				row.Name = name
+				row.Timestamp = ts
+				for _, lp := range m.GetLabel() {
+					row.Tags[lp.GetName()] = lp.GetValue()
+				}
+				row.Tags["pod"] = fw.podName
+				switch mf.GetType() {
+				case io_prometheus_client.MetricType_COUNTER:
+					row.Value = m.GetCounter().GetValue()
+				case io_prometheus_client.MetricType_GAUGE:
+					row.Value = m.GetGauge().GetValue()
+				}
+				r.resultCh <- row
+			}
+		}
+
+		return nil
+	}
+	bo := backoff.NewExponentialBackOff()
+	bo.MaxElapsedTime = 20 * time.Second
+
+	return func() error { return backoff.Retry(op, bo) }
+}
+
+func (r *prometheusRecorderImpl) startPortForward() error {
+	ctx := context.TODO()
```

Comment:
Should this be a `TODO` or just `context.Background()`?

Reply:
now that `Start` has a context I should just pass that through, will fix.